src/epoch/garbage.rs

```diff
@@ -0,0 +1,210 @@
+        let size = mem::size_of::<T>();
+        // We need not drop empty buffers.
+        if size > 0 {
+            // Simply push it to the vector of garbage.
+            self.0.push(Item {
+                ptr: elem as *mut u8,
+                dtor: dtor::<T>,
+            })
+        }
+    }
+
+    /// Get the number of items in this garbage bag.
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Run all destructors of the garbage in the bag.
+    pub unsafe fn collect(&mut self) {
+        // Go over all the garbage, remove it and run the destructors.
+        for item in self.0.drain(..) {
+            (item.dtor)(item.ptr);
+        }
+    }
+}
+
+// Needed because the bags store raw pointers.
+unsafe impl Send for Bag {}
+unsafe impl Sync for Bag {}
+
+/// A thread-local set of garbage bags.
+#[derive(Debug, Default)]
+pub struct Local {
+    /// Garbage added at least one epoch behind the current local epoch.
+    pub old: Bag,
+    /// Garbage added in the current local epoch or earlier.
+    pub cur: Bag,
+    /// Garbage added in the current _global_ epoch.
+    pub new: Bag,
+}
+
+impl Local {
+    /// Insert a new element to be deallocated eventually.
+    pub fn insert<T>(&mut self, elem: *mut T) {
+        self.new.insert(elem)
+    }
+
+    /// Collect one epoch of garbage and rotate the local garbage bags.
+    pub unsafe fn collect(&mut self) {
+        // All the old garbage will be destroyed.
+        let ret = self.old.collect();
+        // Then rotate the garbage (note that `self.old` is now empty).
+        mem::swap(&mut self.old, &mut self.cur);
+        mem::swap(&mut self.cur, &mut self.new);
+
+        ret
+    }
+
+    /// Get the total number of garbage items in the set.
+    pub fn len(&self) -> usize {
+        // Sum the bags.
+        self.old.len() + self.cur.len() + self.new.len()
+    }
+}
+
+/// A concurrent garbage bag.
+///
+/// This is currently based on Treiber's stack. The elements are themselves owned `Bag`s.
+#[derive(Debug, Default)]
+pub struct ConcBag {
+    /// The top node of the stack.
+    head: AtomicPtr<Node>,
+}
+
+/// A node in the Treiber stack of garbage bags.
+#[derive(Debug)]
+struct Node {
+    /// The content of this node.
+    data: Bag,
+    /// The node below.
+    next: AtomicPtr<Node>,
+}
+
+
+impl ConcBag {
+    /// Insert a bag of garbage.
+    pub fn insert(&self, t: Bag) {
+        // Allocate the new node on the heap (so we can get an unbound pointer).
+        let n = Box::into_raw(Box::new(Node {
+            data: t,
+            next: AtomicPtr::new(ptr::null_mut()),
+        }));
+
+        // Spin until the ABA problem is solved.
+        // TODO: Use striping to avoid this from spin multiple times (not as unlikely as you can
+        //       think, especially if several threads constantly creates new garbage).
+        loop {
+            // Load the head.
+            let head = self.head.load(atomic::Ordering::Acquire);
```

Comment:
I see your point now, the first load can be moved to outside of the loop and you can updated it with the result of the cas.